File: backend/app/schemas/device.py
from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class DevicePresetResponse(BaseModel):
    id: UUID
    manufacturer: str
    model: str
    market_name: str
    brand: str
    product_board: str
    hardware: str
    fingerprint: str
    android_versions: List[str]
    supported_abis: List[str]
    screen_specs: Dict[str, Any]

    class Config:
        from_attributes = True


class DeviceRegisterRequest(BaseModel):
    device_identifier: str = Field(min_length=16, max_length=64)
    device_name: str
    manufacturer: Optional[str] = None
    model: Optional[str] = None
    android_version: Optional[str] = None
    sdk_version: Optional[int] = None
    agent_version: Optional[str] = None
    initial_telemetry: Optional[Dict[str, Any]] = None


class DeviceHeartbeatRequest(BaseModel):
    battery_level: Optional[int] = None
    is_charging: Optional[bool] = None
    network_type: Optional[str] = None
    ip_address: Optional[str] = None
    vpn_active: Optional[bool] = None
    developer_options: Optional[bool] = None
    usb_debugging: Optional[bool] = None
    root_detected: Optional[bool] = None
    extra_metrics: Optional[Dict[str, Any]] = None


class DeviceAssignProfileRequest(BaseModel):
    profile_id: Optional[UUID] = None


class DeviceResponse(BaseModel):
    id: UUID
    user_id: UUID
    device_identifier: str
    device_name: str
    manufacturer: Optional[str]
    model: Optional[str]
    android_version: Optional[str]
    sdk_version: Optional[int]
    agent_version: Optional[str]
    status: str
    last_seen: Optional[datetime]
    telemetry: Dict[str, Any]
    assigned_profile_id: Optional[UUID]
    created_at: datetime

    class Config:
        from_attributes = True


class DeviceSyncConfigResponse(BaseModel):
    device_id: UUID
    status: str
    assigned_profile: Optional[Dict[str, Any]]
    effective_applications: List[str]

File: frontend/src/types/device.ts
export interface DeviceTelemetry {
  battery_level: number | null
  is_charging: boolean | null
  network_type: string | null
  ip_address: string | null
  vpn_active: boolean | null
  developer_options: boolean | null
  usb_debugging: boolean | null
  root_detected: boolean | null
  extra_metrics: Record<string, any> | null
}

export interface Device {
  id: string
  user_id: string
  device_identifier: string
  device_name: string
  manufacturer: string | null
  model: string | null
  android_version: string | null
  sdk_version: number | null
  agent_version: string | null
  status: string
  last_seen: string | null
  telemetry: DeviceTelemetry
  assigned_profile_id: string | null
  created_at: string
}

export interface DeviceSyncConfig {
  device_id: string
  status: string
  assigned_profile: Record<string, any> | null
  effective_applications: string[]
}